### PR TITLE
[security][work-in-progress] fuzzing consensus messages

### PR DESCRIPTION
**Pull request by mimoo**
## Abstract

This PR sets up fuzzing of the consensus `event_processor`. Because we're fuzzing at this level, but we don't want to miss on some validation code that happens before, so we copy some of the code from `network.rs` and `chained_bft_smr.rs` and call the `event_processor` functions directly on the node.

This PR also brings a lot of small changes that allow the fuzzer -- which runs with the `fuzzing` cfg flag -- to access useful setup test functions.

## Test Plan

```
cargo test
cargo fuzz run fuzz_consensus_vote -- fuzz/corpus/vote
cargo fuzz run fuzz_consensus_proposal -- fuzz/corpus/proposal
cargo fuzz run fuzz_consensus_timeout -- fuzz/corpus/timeout
```

### CODE REVIEW

File: consensus/src/chained_bft/chained_bft_smr_test.rs [hunk 1]
```diff
@@ -33,18 +34,18 @@ use tokio::runtime;
...
-    validator: Arc<ValidatorVerifier<Ed25519PublicKey>>,
-    peers: Arc<Vec<Author>>,
-    proposer: Vec<Author>,
-    smr_id: usize,
-    smr: ChainedBftSMR<TestPayload>,
-    commit_cb_receiver: mpsc::UnboundedReceiver<LedgerInfoWithSignatures<Ed25519Signature>>,
-    mempool: Arc<MockTransactionManager>,
-    mempool_notif_receiver: mpsc::Receiver<usize>,
-    storage: Arc<MockStorage<TestPayload>>,
+pub struct SMRNode {
```
Comment (gedigi) [inline consensus/src/chained_bft/chained_bft_smr_test.rs]: should this be gated behind cfg-fuzzing?

File: consensus/src/chained_bft/chained_bft_smr_test.rs [hunk 2]
```diff
@@ -140,7 +141,7 @@ impl SMRNode {
         )
     }
 
-    fn start_num_nodes(
+    pub fn start_num_nodes(
```
Comment (gedigi) [inline consensus/src/chained_bft/chained_bft_smr_test.rs]: same here

File: consensus/src/chained_bft/event_processor_test.rs [hunk 3]
```diff
@@ -60,18 +62,23 @@ use types::{
...
-    author: Author,
-    block_store: Arc<BlockStore<TestPayload>>,
-    event_processor: EventProcessor<TestPayload>,
-    new_rounds_receiver: channel::Receiver<NewRoundEvent>,
-    storage: Arc<MockStorage<TestPayload>>,
-    signer: ValidatorSigner<Ed25519PrivateKey>,
-    proposer_author: Author,
-    peers: Arc<Vec<Author>>,
-    pacemaker: Arc<dyn Pacemaker>,
+pub struct NodeSetup {
```
Comment (gedigi) [inline consensus/src/chained_bft/event_processor_test.rs]: cfg-fuzzing?

File: consensus/src/chained_bft/event_processor_test.rs [hunk 4]
```diff
@@ -127,10 +142,11 @@ impl NodeSetup {
         Arc::new(RotatingProposer::new(vec![author], 1))
     }
 
-    fn create_nodes(
+    pub fn create_nodes(
```
Comment (gedigi) [inline consensus/src/chained_bft/event_processor_test.rs]: cfg-fuzzing

File: consensus/fuzz/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        });
+    }
+}
+
+#[allow(dead_code)]
+fn fuzz_single_file() {
+    // fuzz input
+    let mut buf = [0; 1000];
+    let mut f =
+        File::open("/Users/davidwg/libra_internal/libra/consensus/fuzz/corpus/proposal.bin")
```
Comment (huitseeker) [inline consensus/fuzz/src/main.rs]: Ouch!

**Comment by gedigi** [comment]
the bin files are the base corpus for the fuzzer. where do you recommend storing them?

**Comment by mimoo** [comment]
btw this is unpolished, don't accept. I need to discuss the approach with consensus, as well as integrate this into @sunshowers framework

**Comment by huitseeker** [comment]
> the bin files are the base corpus for the fuzzer. where do you recommend storing them?

I get what they are meant to be. Yet in practice, i suspect they are a much simpler product. How do they co-evolve with our (*rapidly* changing) serialization format. Is this maintainable? 

To wit:
```
fga@fga-ubuntu-PC0UYPZ9➜tmp/calibra/consensus(pr/588✗)» hexdump -C fuzz/corpus/vote/vote.bin                                                                                                [15:20:19]
00000000  12 d6 02 0a 20 11 41 2b  c4 bc f4 35 d3 19 bf ee  |.... .A+...5....|
00000010  35 7f ef 95 23 9e ea 46  ce 9d 5d a4 51 0e f1 62  |5...#..F..].Q..b|
00000020  18 e1 e0 f1 95 12 20 41  43 43 55 4d 55 4c 41 54  |...... ACCUMULAT|
00000030  4f 52 5f 50 4c 41 43 45  48 4f 4c 44 45 52 5f 48  |OR_PLACEHOLDER_H|
00000040  41 53 48 00 00 00 00 20  01 2a 20 01 00 00 00 00  |ASH.... .* .....|
00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000060  00 00 00 00 00 00 00 00  00 00 00 32 66 12 20 00  |...........2f. .|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000080  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 1a  |................|
00000090  20 1a b5 ff 67 2b c4 07  36 fd 4c 13 07 d1 49 42  | ...g+..6.L...IB|
000000a0  74 23 01 27 f7 1d b4 1b  84 7d b3 ad 74 2e c0 b9  |t#.'.....}..t...|
000000b0  ba 22 20 00 00 00 00 00  00 00 00 00 00 00 00 00  |." .............|
000000c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000d0  00 00 00 3a 40 aa be 08  d2 af b4 96 d7 f7 4f 56  |...:@.........OV|
000000e0  6c 33 ef e5 c2 e3 f4 96  b8 ba 2d 69 6e 77 ef e1  |l3........-inw..|
000000f0  fe 90 a7 2f d9 68 d0 2a  ce d5 e4 a4 9b fc a7 69  |.../.h.*.......i|
00000100  19 c5 87 46 0f 11 80 5d  17 0c 65 48 ef e8 8e 4d  |...F...]..eH...M|
00000110  e0 f9 cb 39 0c 42 20 47  45 4e 45 53 49 53 5f 42  |...9.B GENESIS_B|
00000120  4c 4f 43 4b 5f 49 44 00  00 00 00 00 00 00 00 00  |LOCK_ID.........|
00000130  00 00 00 00 00 00 00 52  20 47 45 4e 45 53 49 53  |.......R GENESIS|
00000140  5f 42 4c 4f 43 4b 5f 49  44 00 00 00 00 00 00 00  |_BLOCK_ID.......|
00000150  00 00 00 00 00 00 00 00  00                       |.........|
00000159
```

**Comment by gedigi** [comment]
makes sense. we should provide a corpus generator a-la @sunshowers's approach

**Comment by mimoo** [comment]
@huitseeker good catch. I'm used to write stuff for protocols that are already written in stone :D

**Comment by yevgenypats** [comment]
Hi @mimoo @gedigi @huitseeker  I think my PR might relevant as well and maybe we can integrate the new fuzzer in this PR to continuous fuzzing too. I don't know who is right person to review my PR so I link to it here
https://github.com/libra/libra/pull/518

**Comment by mimoo** [comment]
This has been taken care of in a different PR.

**Pull request closed without merging**